### PR TITLE
BPM-38 Create category endpoint

### PR DESCRIPTION
**Pull request by SpraylnlPray**
created the crud functionality for the categories endpoint

### CODE REVIEW

File: src/main/java/com/ioet/bpm/skills/boundaries/CategoryController.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+@RequestMapping("/categories")
+public class CategoryController {
+
+    private final CategoryRepository categoryRepository;
+
+    public CategoryController(CategoryRepository categoryRepository) {
+        this.categoryRepository = categoryRepository;
+    }
+
+    @ApiOperation(value = "Return a list of all categorys", response = Category.class, responseContainer = "List")
```
Comment (enriquezrene) [inline src/main/java/com/ioet/bpm/skills/boundaries/CategoryController.java]: categories or categorys?

File: src/main/java/com/ioet/bpm/skills/boundaries/CategoryController.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        return categoryOptional.map(
+                category -> new ResponseEntity<>(category, HttpStatus.OK))
+                .orElseGet(() -> new ResponseEntity<>(HttpStatus.NOT_FOUND));
+    }
+
+    @ApiOperation(value = "Create a new category", response = Category.class)
+    @ApiResponses(value = {
+            @ApiResponse(code = 201, message = "Category successfully created")
+    })
+    @PostMapping(produces = "application/json")    public ResponseEntity<Category> createCategory(@RequestBody Category category) {
```
Comment (enriquezrene) [inline src/main/java/com/ioet/bpm/skills/boundaries/CategoryController.java]: can we add a break line after the @PostMapping annotation?

File: src/main/java/com/ioet/bpm/skills/boundaries/SkillController.java [hunk 3]
```diff
@@ -53,16 +57,11 @@ public SkillController(SkillRepository skillRepository) {
     @PutMapping("/{id}")
     public ResponseEntity<Skill> updateSkill(@PathVariable(value = "id") String skillId,
                                              @Valid @RequestBody Skill skillDetails) {
-
-        Optional<Skill> skillOptional = skillRepository.findById(skillId);
-        if (!skillOptional.isPresent()) {
+        try{
+            return new ResponseEntity<>(skillService.save(skillId,skillDetails),  HttpStatus.OK);
+        }
+        catch (EntityNotFoundException ex){
```
Comment (enriquezrene) [inline src/main/java/com/ioet/bpm/skills/boundaries/SkillController.java]: Please get rid of this exception, check the comments I added in @juanultimate's PR

File: src/main/java/com/ioet/bpm/skills/domain/Category.java [hunk 4]
```diff
@@ -0,0 +1,19 @@
+package com.ioet.bpm.skills.domain;
+
+import lombok.AllArgsConstructor;
+import lombok.Data;
+
+import javax.validation.constraints.NotBlank;
+
+@Data
+@AllArgsConstructor
+public class Category {
```
Comment (enriquezrene) [inline src/main/java/com/ioet/bpm/skills/domain/Category.java]: Is there a reason to don't have this field marked as private?

File: src/main/java/com/ioet/bpm/skills/domain/Category.java [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+    @NotBlank
+    String name;
+    @NotBlank
+    Double businessValue;
+    @NotBlank
+    Double predictiveValue;
+    @NotBlank
+    String Id;
```
Comment (enriquezrene) [inline src/main/java/com/ioet/bpm/skills/domain/Category.java]: Don't we need an additional annotation to mark this field as Pk?

File: src/main/java/com/ioet/bpm/skills/boundaries/SkillController.java (hunk 3 above)
Comment (SpraylnlPray) [reply src/main/java/com/ioet/bpm/skills/boundaries/SkillController.java]: Since this is part of the code @juanultimate wrote, I'll wait until his changes are in master and then integrate those into my code.

File: src/main/java/com/ioet/bpm/skills/boundaries/CategoryController.java (hunk 1 above)
Comment (SpraylnlPray) [reply src/main/java/com/ioet/bpm/skills/boundaries/CategoryController.java]: Done

File: src/main/java/com/ioet/bpm/skills/boundaries/CategoryController.java (hunk 2 above)
Comment (SpraylnlPray) [reply src/main/java/com/ioet/bpm/skills/boundaries/CategoryController.java]: Done

File: src/main/java/com/ioet/bpm/skills/domain/Category.java (hunk 5 above)
Comment (SpraylnlPray) [reply src/main/java/com/ioet/bpm/skills/domain/Category.java]: Done

File: src/main/java/com/ioet/bpm/skills/domain/Category.java (hunk 4 above)
Comment (SpraylnlPray) [reply src/main/java/com/ioet/bpm/skills/domain/Category.java]: Done

File: src/main/java/com/ioet/bpm/skills/boundaries/SkillController.java (hunk 3 above)
Comment (SpraylnlPray) [reply src/main/java/com/ioet/bpm/skills/boundaries/SkillController.java]: Done

File: src/main/java/com/ioet/bpm/skills/boundaries/CategoryController.java [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+    })
+    @DeleteMapping(path = "/{id}", produces = "application/json")
+    public ResponseEntity<?> deleteCategory(@PathVariable(value = "id") String categoryId) {
+        Optional<Category> category = categoryRepository.findById(categoryId);
+        if (!category.isPresent()) {
+            return new ResponseEntity<>(HttpStatus.NOT_FOUND);
+        }
+
+        categoryRepository.delete(category.get());
+        return new ResponseEntity<>(HttpStatus.OK);
```
Comment (enriquezrene) [inline src/main/java/com/ioet/bpm/skills/boundaries/CategoryController.java]: You can make it easier to read in this way:

```
if (category.isPresent()) {
 categoryRepository.delete(category.get());
 return new ResponseEntity<>(HttpStatus.OK);
}  else  {
  return new ResponseEntity<>(HttpStatus.NOT_FOUND);
}
```

Comment (enriquezrene) [reply src/main/java/com/ioet/bpm/skills/boundaries/CategoryController.java]: For delete you need to return NOT_CONTENT instead of OK

File: src/main/java/com/ioet/bpm/skills/boundaries/SkillController.java [hunk 7]
```diff
@@ -51,18 +51,15 @@ public SkillController(SkillRepository skillRepository) {
...
+            return new ResponseEntity<>(skillSaved, HttpStatus.OK);
         }
-
-        Skill skill = skillOptional.get();
-        skill.setName(skillDetails.getName());
-        skill.setDescription(skillDetails.getDescription());
-        skillRepository.save(skill);
-        return new ResponseEntity<>(skill, HttpStatus.OK);
+        
+        return new ResponseEntity<>("Skill with id " + skillId + " not updated.", HttpStatus.NOT_FOUND);
```
Comment (enriquezrene) [inline src/main/java/com/ioet/bpm/skills/boundaries/SkillController.java]: I'd say:
```
"Skill with id " + skillId + " does not exist."
```

File: .gitignore [hunk 8]
```diff
@@ -17,10 +17,12 @@
...
 
 ### NetBeans ###
 /nbproject/private/
 /nbbuild/
 /dist/
 /nbdist/
-/.nb-gradle/
\ No newline at end of file
+/.nb-gradle/
+
```
Comment (enriquezrene) [inline .gitignore]: You might want to include the bin folder here.
You are adding this file which is absolutely unnecessary:
bin/main/application.properties

File: README.md [hunk 9]
```diff
@@ -16,5 +16,8 @@ The project has integrated a docker plugin so you can generate a docker image us
...
 ```
 
-Don't forget to pass the `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` to make it work locally.
-For any other environment the credentials should be provided by the CI server.
\ No newline at end of file
+##Developer
+
+Don't forget to pass the `AWS_ACCESS_KEY_ID`, `AWS_SECRET_ACCESS_KEY` and `AWS_REGION` to make it work locally.
+For any other environment the credentials should be provided by the CI server.
+
```
Comment (enriquezrene) [inline README.md]: How do you pass them?
Can you please improve the README, you can feel inspired by following this file: https://github.com/ioet/bpm-people-api

File: bin/main/application.properties [hunk 10]
```diff
@@ -0,0 +1,4 @@
+spring.application.name=skills-service
```
Comment (enriquezrene) [inline bin/main/application.properties]: is this file needed somehow?

File: src/main/java/com/ioet/bpm/skills/boundaries/SkillController.java [hunk 11]
```diff
@@ -1,65 +1,105 @@
...
+        if (!skillOptional.isPresent()) {
+            return new ResponseEntity<>(HttpStatus.NOT_FOUND);
         }
 
-        return new ResponseEntity<>("Skill with id " + skillId + " does not exist.", HttpStatus.NOT_FOUND);
+        Skill skill = skillOptional.get();
+        skill.setName(skillDetails.getName());
+        skill.setBusinessValue(skillDetails.getBusinessValue());
+        skill.setPredictiveValue(skillDetails.getPredictiveValue());
+        skillRepository.save(skill);
```
Comment (enriquezrene) [inline src/main/java/com/ioet/bpm/skills/boundaries/SkillController.java]: You can do instead:

skillDetails.setId(skill.getId());

and then save skillDetails.

File: src/main/java/com/ioet/bpm/skills/domain/Skill.java [hunk 12]
```diff
@@ -1,36 +1,36 @@
...
 @Data
 @DynamoDBTable(tableName = "skills_skill")
 public class Skill {
 
+    @Id
     @DynamoDBHashKey
     @DynamoDBAutoGeneratedKey
     private String id;
 
-    @NotBlank
```
Comment (enriquezrene) [inline src/main/java/com/ioet/bpm/skills/domain/Skill.java]: Do you want skills with a null name?

File: bin/main/application.properties (hunk 10 above)
Comment (SpraylnlPray) [reply bin/main/application.properties]: I deleted the file

File: src/main/java/com/ioet/bpm/skills/domain/Skill.java (hunk 12 above)
Comment (SpraylnlPray) [reply src/main/java/com/ioet/bpm/skills/domain/Skill.java]: This is the same in master, should i change it anyways?

File: .gitignore (hunk 8 above)
Comment (SpraylnlPray) [reply .gitignore]: the bin folder is already in the git ignore i guess

File: src/main/java/com/ioet/bpm/skills/boundaries/SkillController.java [hunk 13]
```diff
@@ -87,18 +90,16 @@ public SkillController(SkillRepository skillRepository, CategoryRepository categ
...
+
+        Optional<Skill> skillFound = skillRepository.findById(skillId);
+        if (!skillFound.isPresent()) {
+            return new ResponseEntity<>(HttpStatus.NOT_FOUND);
         }
-        return new ResponseEntity<>(HttpStatus.NOT_FOUND);
+
+        skillToUpdate.setId(skillFound.get().getId());
+        Skill updatedSkill = skillRepository.save(skillToUpdate);
+        return new ResponseEntity<>(updatedSkill, HttpStatus.OK);
```
Comment (enriquezrene) [inline src/main/java/com/ioet/bpm/skills/boundaries/SkillController.java]: You can make readable in this way:
```
if (skillFound.isPresent()) {
        return new ResponseEntity<>(HttpStatus.NOT_FOUND);
         skillToUpdate.setId(skillFound.get().getId());
        Skill updatedSkill = skillRepository.save(skillToUpdate);
        return new ResponseEntity<>(updatedSkill, HttpStatus.OK);

        }
            return new ResponseEntity<>(HttpStatus.NOT_FOUND);
```

File: src/main/java/com/ioet/bpm/skills/domain/converters/CategoryConverter.java [hunk 14]
```diff
@@ -0,0 +1,54 @@
+package com.ioet.bpm.skills.domain.converters;
+
+
+import com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBTypeConverter;
+import com.ioet.bpm.skills.domain.Category;
+import com.ioet.bpm.skills.domain.converters.exception.ConverterException;
+
+public class CategoryConverter implements DynamoDBTypeConverter<String, Category> {
```
Comment (enriquezrene) [inline src/main/java/com/ioet/bpm/skills/domain/converters/CategoryConverter.java]: is this class still needed? 
I think it is not.

File: src/main/java/com/ioet/bpm/skills/domain/converters/exception/ConverterException.java [hunk 15]
```diff
@@ -0,0 +1,15 @@
+package com.ioet.bpm.skills.domain.converters.exception;
```
Comment (enriquezrene) [inline src/main/java/com/ioet/bpm/skills/domain/converters/exception/ConverterException.java]: This class should not be needed either.

File: src/test/java/com/ioet/bpm/skills/domain/converters/CategoryConverterTest.java [hunk 16]
```diff
@@ -0,0 +1,22 @@
+package com.ioet.bpm.skills.domain.converters;
+
+import com.ioet.bpm.skills.domain.Category;
+import org.junit.Test;
+
+import static org.junit.Assert.*;
+
+public class CategoryConverterTest {
```
Comment (enriquezrene) [inline src/test/java/com/ioet/bpm/skills/domain/converters/CategoryConverterTest.java]: this class is not needed.

File: src/main/java/com/ioet/bpm/skills/boundaries/CategoryController.java [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+    @ApiResponses(value = {
+            @ApiResponse(code = 204, message = "Category successfully deleted"),
+            @ApiResponse(code = 404, message = "The category to delete was not found")
+    })
+    @DeleteMapping(path = "/{id}", produces = "application/json")
+    public ResponseEntity<?> deleteCategory(@PathVariable(value = "id") String categoryId) {
+        Optional<Category> category = categoryRepository.findById(categoryId);
+        if (category.isPresent()) {
+            categoryRepository.delete(category.get());
+            return new ResponseEntity<>(HttpStatus.OK);
```
Comment (oemel09) [inline src/main/java/com/ioet/bpm/skills/boundaries/CategoryController.java]: Instead of 200 OK you can return a 204 No Content

File: src/main/java/com/ioet/bpm/skills/domain/converters/CategoryConverter.java (hunk 14 above)
Comment (oemel09) [reply src/main/java/com/ioet/bpm/skills/domain/converters/CategoryConverter.java]: Probably you can delete the whole converters package as well as the domain package in the test folder

File: src/main/java/com/ioet/bpm/skills/boundaries/SkillController.java [hunk 18]
```diff
@@ -87,17 +90,15 @@ public SkillController(SkillRepository skillRepository, CategoryRepository categ
...
-            return new ResponseEntity<>( "categoryId not found", HttpStatus.UNPROCESSABLE_ENTITY);
+    public ResponseEntity<?> updateSkill(@PathVariable(value = "id") String skillId,
+                                             @Valid @RequestBody Skill skillToUpdate) {
+
+        Optional<Skill> skillFound = skillRepository.findById(skillId);
+
+        if (skillFound.isPresent()) {
+            skillToUpdate.setId(skillFound.get().getId());
+            Skill updatedSkill = skillRepository.save(skillToUpdate);
+            return new ResponseEntity<>(updatedSkill, HttpStatus.OK);
```
[the remark this answers is not in the document]
Comment (SpraylnlPray) [reply src/main/java/com/ioet/bpm/skills/boundaries/SkillController.java]: oops
